overwatch-derive/src/lib.rs

```diff
@@ -1077,3 +1177,48 @@ fn generate_as_service_id_impl(fields: &Punctuated<Field, Comma>) -> proc_macro2
         #(#impl_blocks)*
     }
 }
+
+fn create_finished_signal_channels_from_amount(amount: usize) -> proc_macro2::TokenStream {
+    quote! {
+        let channels = (0..#amount).map(|_| { ::overwatch::utils::finished_signal::channel() });
+        let (mut senders, receivers): (Vec<_>, Vec<_>) = channels.into_iter().unzip();
+    }
+}
+
+fn create_finished_signal_channels_from_variable(variable: &Ident) -> proc_macro2::TokenStream {
+    quote! {
+        let channels = (0..#variable).map(|_| { ::overwatch::utils::finished_signal::channel() });
+        let (mut senders, receivers): (Vec<_>, Vec<_>) = channels.into_iter().unzip();
+    }
+}
+
+fn await_finished_signal_receivers() -> proc_macro2::TokenStream {
+    quote! {
+        for mut receiver in receivers {
+            receiver.await.map_err(|error| {
+                let dyn_error: ::overwatch::DynError = Box::new(error);
+                ::overwatch::overwatch::Error::from(dyn_error)
+            })?;
```

Comment:
Both starting and stopping are done in a sequential pattern. We could do it completely asynchronous unless there is a reason to do it in order. In this case I think is ok if we want to keep the ordering of starting and stopping for example. But we should document it. 
Otherwise all at the same time should be fine as well.

Reply:
Lets keep sequential and just add documentation that the order of the start/stop services is up to the caller.